feat(protocol/deposit-flow): Low level documentation of bedrock deposit flow (#733)
ethereum-optimism/community-hub: src/docs/protocol/deposit-flow.md

---
title: Deposit Flow
lang: en-US
---

## Introduction

> In Optimism terminology *deposit* refers to any transaction that goes from L1 to L2.

Reviewer: ```suggestion
In Optimism terminology, the term "*deposit transaction*" refers to any L2 transaction that is triggered by a transaction or event on L1.
```

A deposit transaction may or may not have assets (ETH, tokens, etc.) attached to it.

The process is somewhat similar to [the way most networking stacks work](https://en.wikipedia.org/wiki/Encapsulation_(networking)).
Information is encapsulated in lower layer packets on the sending side, and then retrieved and used by those layers on the receiving side while going up the stack to the receiving application.

> ![Overall process](../../assets/docs/protocol/deposit-flow/overall-process.svg)

Reviewer: Same comment as below re: maybe worth noting OptimismPortal direct access

Author: I prefer to leave the illustration as simple as possible (which is still somewhat complicated).

## L1 Processing

> 1. An L1 entity, either a smart contract or an externally owned account (EOA), sends a deposit transaction to [`L1CrossDomainMessenger`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/contracts/L1/L1CrossDomainMessenger.sol), using [`sendMessage`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol#L259). 

Reviewer: Maybe worth noting that an L1 account can call the `OptimismPortal` contract directly but will not be able to benefit from certain automatic safety protections and devex improvements.

  This function accepts three parameters:

   - `_target`, target address on L2.

>    - `_message`, the L2 transaction's calldata, [formatted as per the ABI](https://docs.soliditylang.org/en/v0.8.19/abi-spec.html).

Reviewer: ```suggestion
   - `_message`, the L2 transaction's calldata, formatted as per the [ABI](https://docs.soliditylang.org/en/v0.8.19/abi-spec.html) of the target account.
```

>    - `_minGasLimit`, the gas limit required for the transaction on L2. 

Reviewer: ```suggestion
   - `_minGasLimit`, the minimum gas limit allowed for the transaction on L2. Note that this is a *minimum* and the actual amount provided on L2 may be higher (but never lower) than the specified gas limit.
```

   ```

1. Send a greeting change as a deposit.
   Use these commands:

   ```sh
   L1_RPC=<URL to Goerli>
   L1_CROSS_DOM_COMM=0x5086d1eef304eb5284a0f6720f79403b4e9be294
   FUNC="sendMessage(address,bytes,uint32)"
   CALLDATA=`cast calldata "setGreeting(string)" "testing"`
   cast send --rpc-url $L1_RPC --private-key $PRIV_KEY $L1_CROSS_DOM_COMM $FUNC $GREETER $CALLDATA 10000000
   ```

   The transaction will be successful on L1, but then emit a fail event on L2.

1. The next step is to find the hash of the failed relay.
   The easiest way to do this is to look in [the internal transactions of the destination contract](https://goerli-optimism.etherscan.io/address/0x26A145eccDf258688C763726a8Ab2aced898ADe1#internaltx), and select the latest one that appears as a failure.
   It should be a call to L2CrossDomainMessenger at address `0x420...007`. This is the call you need to replay.

   If the latest internal transaction is a success, it probably means your transaction hasn't relayed yet. When until it is, that may take a few minutes.

1. Get the transaction information using Foundry.

   ```sh
   TX_HASH=<transaction hash from Etherscan>

>    L2_CROSS_DOM_COMM=0x4200000000000000000000000000000000000007

Reviewer: ```suggestion
   L2XDM_ADDRESS=0x4200000000000000000000000000000000000007
```

Reviewer: Maybe. But "COMM" doesn't fit here, it's `L2CrossDomainMessenger` and not `L2CrossDomainCommunicator`

   REPLAY_DATA=`cast tx $TX_HASH input`
   ```

1. Call `startChanges()` to allow changes using this Foundry command:

   ```sh
   cast send --private-key $PRIV_KEY $GREETER "startChanges()"
   ```

   ::: warning Don't do this prematurely

   If you call `startChanges()` too early, it will happen from the message is relayed to L2, and then the initial deposit will be successful and there will be no need to replay it.

   :::

1. Verify that `getStatus()` returns true, meaning changes are not allowed, and see the value of `greet()`. 
   Foundry returns true as one.

   ```sh
   cast call $GREETER "greet()" | cast --to-ascii ; cast call $GREETER "getStatus()"
   ```

1. Actually send the replay transaction.

   ```sh   
   cast send --private-key $PRIV_KEY --gas-limit 10000000 $L2_CROSS_DOM_COMM $REPLAY_DATA 
   ```

   ::: details Why do we need to specify the gas limit?
   
   The gas estimation mechanism tries to find the minimum gas limit at which the transaction would be successful. 

>    However, the L2 cross domain messenger does not revert when a replay fails due to low gas limit, it just emits a failure message. 

Reviewer: Nit let's use `L2CrossDomainMessenger` (with backticks) instead of "L2 cross domain messenger".

## Denial of service (DoS) prevention

As with all other L1 transactions, the L1 costs of a deposit are borne by the transaction's originator.
However, the L2 processing of the transaction is performed by the Optimism nodes.
If there were no cost attached, an attacker could be able to submit a transaction that had high costs of run on L2, and that way perform a denial of service attack.

To avoid this DoS vector, [`depositTransaction`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/contracts/L1/OptimismPortal.sol#L434), and the functions that call it, require a gas limit parameter.
[This gas limit is encoded into the `TransactionDeposited`(https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/contracts/L1/OptimismPortal.sol#L472-L482) event], and used as the gas limit for the user deposit transaction on L2.

This L2 gas is paid for by burning L1 gas [here](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/contracts/L1/ResourceMetering.sol#L162). 

## Replaying messages

> Deposits are transactions, and as such can fail due to several reasons:

Reviewer 2: Personally think it flows better keeping it simple:

```suggestion
Deposits transactions can fail due to several reasons:
```

- Not enough gas provided.
- The state on L2 does not allow the transaction to be successful.

> It is possible to replay a failed deposit, possibly with more gas, 

Author: It is, see below on exactly how to get it working.
